File: sdk/python/src/soulfire/connection.py
from __future__ import annotations

from dataclasses import dataclass
from types import MappingProxyType
from typing import Final

from .plugin_api_pb2 import PluginApiDescriptor
from .sdk_pb2 import SdkApiVersion, SdkHandshakeResponse, SdkIdentity

SDK_VERSION: Final = "2.10.1"
SDK_API_VERSION: Final = SdkApiVersion(major=1, minor=0, patch=0)


class SoulFireCompatibilityError(RuntimeError):
    """The server and SDK cannot safely communicate."""


class SoulFireCapabilityError(RuntimeError):
    def __init__(self, capability: str) -> None:
        super().__init__(f"SoulFire capability is unavailable: {capability}")
        self.capability = capability


@dataclass(frozen=True, slots=True)
class RequiredPlugin:
    plugin_id: str
    version_range: str | None = None


@dataclass(frozen=True, slots=True)
class ServerMetadata:
    id: str
    version: str
    commit_hash: str
    branch_name: str
    api_version: SdkApiVersion
    minecraft_version: str
    supported_minecraft_versions: tuple[str, ...]
    transports: tuple[int, ...]


class CapabilitySet:
    __slots__ = ("_revisions",)

    def __init__(self, response: SdkHandshakeResponse) -> None:
        self._revisions = MappingProxyType(
            {capability.id: capability.revision for capability in response.capabilities}
        )

    def supports(self, capability: str, *, minimum_revision: int = 1) -> bool:
        return self._revisions.get(capability, 0) >= minimum_revision

    def require(self, capability: str, *, minimum_revision: int = 1) -> None:
        if not self.supports(capability, minimum_revision=minimum_revision):
            raise SoulFireCapabilityError(capability)

    def revision(self, capability: str) -> int | None:
        return self._revisions.get(capability)

    @property
    def revisions(self) -> MappingProxyType[str, int]:
        return self._revisions


@dataclass(frozen=True, slots=True)
class ConnectionMetadata:
    server: ServerMetadata
    identity: SdkIdentity
    capabilities: CapabilitySet
    plugins: tuple[PluginApiDescriptor, ...]
    limits: MappingProxyType[str, int]

    @classmethod
    def from_response(cls, response: SdkHandshakeResponse) -> ConnectionMetadata:
        if not response.HasField("api_version") or not response.HasField("identity"):
            raise SoulFireCompatibilityError("SoulFire returned an incomplete SDK handshake")
        return cls(
            server=ServerMetadata(
                id=response.server_id,
                version=response.soulfire_version,
                commit_hash=response.commit_hash,
                branch_name=response.branch_name,
                api_version=response.api_version,
                minecraft_version=response.native_minecraft_version,
                supported_minecraft_versions=tuple(response.supported_minecraft_versions),
                transports=tuple(response.transports),
            ),
            identity=response.identity,
            capabilities=CapabilitySet(response),
            plugins=tuple(response.plugins),
            limits=MappingProxyType({limit.id: limit.value for limit in response.limits}),
        )

File: sdk/typescript/src/connection.ts
import type {
  PluginApiDescriptor,
} from "./generated/soulfire/plugin_api_pb.js";
import type {
  SdkApiVersion,
  SdkHandshakeResponse,
  SdkIdentity,
  SdkTransport,
} from "./generated/soulfire/sdk_pb.js";

export const SDK_API_VERSION = {
  major: 1,
  minor: 0,
  patch: 0,
} as const;

export const SDK_VERSION = "2.10.1";

export interface ServerMetadata {
  readonly id: string;
  readonly version: string;
  readonly commitHash: string;
  readonly branchName: string;
  readonly apiVersion: Readonly<SdkApiVersion>;
  readonly minecraftVersion: string;
  readonly supportedMinecraftVersions: readonly string[];
  readonly transports: readonly SdkTransport[];
}

export class SoulFireCompatibilityError extends Error {
  public constructor(
    message: string,
    public override readonly cause?: unknown,
  ) {
    super(message, cause === undefined ? undefined : { cause });
    this.name = "SoulFireCompatibilityError";
  }
}

export class SoulFireCapabilityError extends Error {
  public constructor(public readonly capability: string) {
    super(`SoulFire capability is unavailable: ${capability}`);
    this.name = "SoulFireCapabilityError";
  }
}

export class CapabilitySet {
  readonly #revisions: ReadonlyMap<string, number>;

  public constructor(handshake: SdkHandshakeResponse) {
    this.#revisions = new Map(
      handshake.capabilities.map(({ id, revision }) => [id, revision]),
    );
  }

  public supports(capability: string, minimumRevision = 1): boolean {
    return (this.#revisions.get(capability) ?? 0) >= minimumRevision;
  }

  public require(capability: string, minimumRevision = 1): void {
    if (!this.supports(capability, minimumRevision)) {
      throw new SoulFireCapabilityError(capability);
    }
  }

  public revision(capability: string): number | undefined {
    return this.#revisions.get(capability);
  }

  public entries(): ReadonlyMap<string, number> {
    return this.#revisions;
  }
}

export interface ConnectionMetadata {
  readonly server: ServerMetadata;
  readonly identity: Readonly<SdkIdentity>;
  readonly capabilities: CapabilitySet;
  readonly plugins: readonly PluginApiDescriptor[];
  readonly limits: ReadonlyMap<string, bigint>;
}

export function connectionMetadata(
  handshake: SdkHandshakeResponse,
): ConnectionMetadata {
  if (
    handshake.apiVersion === undefined
    || handshake.identity === undefined
  ) {
    throw new SoulFireCompatibilityError(
      "SoulFire returned an incomplete SDK handshake",
    );
  }
  return {
    server: {
      id: handshake.serverId,
      version: handshake.soulfireVersion,
      commitHash: handshake.commitHash,
      branchName: handshake.branchName,
      apiVersion: handshake.apiVersion,
      minecraftVersion: handshake.nativeMinecraftVersion,
      supportedMinecraftVersions: handshake.supportedMinecraftVersions,
      transports: handshake.transports,
    },
    identity: handshake.identity,
    capabilities: new CapabilitySet(handshake),
    plugins: handshake.plugins,
    limits: new Map(handshake.limits.map(({ id, value }) => [id, value])),
  };
}
